core/Enclave/key_operation.cpp

```diff
@@ -0,0 +1,2047 @@
+        goto out;
+    }
+    if (EVP_MD_CTX_init(mdctx) != 1)
+    {
+        log_d("ecall sm2_sign EVP_MD_CTX initialize failed.\n");
+        goto out;
+    }
+
+    // set sm2 evp pkey
+    if (EVP_PKEY_set_alias_type(evpkey, EVP_PKEY_SM2) != 1)
+    {
+        log_d("ecall sm2_sign fail to modify the evpkey to use SM2\n");
+        goto out;
+    }
+    pkey_ctx = EVP_PKEY_CTX_new(evpkey, NULL);
+    if (pkey_ctx == NULL)
+    {
+        log_d("ecall sm2_sign fail to create a EVP_PKEY_CTX\n");
+        goto out;
+    }
+    if (EVP_PKEY_CTX_set1_id(pkey_ctx, g_default_userid, g_default_userid_len) != 1)
+    {
+        log_d("ecall sm2_sign fail to set sm2_user_id to the EVP_PKEY_CTX\n");
+        goto out;
+    }
+    EVP_MD_CTX_set_pkey_ctx(mdctx, pkey_ctx);
+
+    if (EVP_DigestSignInit(mdctx, &pkey_ctx, digestMode, nullptr, evpkey) != 1)
+    {
+        log_d("ecall sm2_sign EVP_DigestSignInit failed.\n");
+        goto out;
+    }
+
+    if (EVP_DigestUpdate(mdctx, data->data, data->datalen) != 1)
+    {
+        log_d("ecall sm2_sign EVP_DigestSignUpdate failed.\n");
+        goto out;
+    }
+    if (EVP_DigestSignFinal(mdctx, NULL, &temp_signature_size) != 1)
+    {
+        log_d("ecall sm2_sign EVP_DigestSignFinal1 failed.\n");
+        goto out;
+    }
+    if (EVP_DigestSignFinal(mdctx, signature->data, &temp_signature_size) != 1)
+    {
+        log_d("ecall sm2_sign EVP_DigestSignFinal failed.\n");
+        goto out;
+    }
+    // return the exact length
+    signature->datalen = (uint32_t)temp_signature_size;
+
+    ret = SGX_SUCCESS;
+
+out:
+    if (ec_key)
+        EC_KEY_free(ec_key);
+    if (bio)
+        BIO_free(bio);
+    if (evpkey)
+        EVP_PKEY_free(evpkey);
+    if (mdctx)
+        EVP_MD_CTX_free(mdctx);
+    if (pkey_ctx)
+        EVP_PKEY_CTX_free(pkey_ctx);
+
+    memset_s(ec_keypair, cmk->keybloblen, 0, cmk->keybloblen);
+    SAFE_FREE(ec_keypair);
+
+    return ret;
+}
+
+/**
+ * @brief make sm2 verify with the designated digest mode
+ * digest mode is optional
+ * running in enclave
+ * @param cmk_blob cipher block for storing keys
+ * @param data data to be signed
+ * @param signature generated signature
+ * @param result match result
+ * @return sgx_status_t
+ */
+sgx_status_t ehsm_sm2_verify(const ehsm_keyblob_t *cmk,
+                             const ehsm_data_t *data,
+                             const ehsm_data_t *signature,
+                             bool *result)
+{
+    sgx_status_t ret = SGX_ERROR_UNEXPECTED;
+
+    uint8_t *ec_keypair = NULL;
+    BIO *bio = NULL;
+    EC_KEY *ec_key = NULL;
+    EVP_PKEY *evpkey = NULL;
+    EVP_MD_CTX *mdctx = NULL;
+    EVP_PKEY_CTX *pkey_ctx = NULL;
+
+    const EVP_MD *digestMode = GetDigestMode(cmk->metadata.digest_mode);
+    if (digestMode == NULL)
+    {
+        log_d("ecall sm2_verify digestMode error.\n");
+        ret = SGX_ERROR_INVALID_PARAMETER;
+        goto out;
+    }
+
+    ec_keypair = (uint8_t *)malloc(cmk->keybloblen);
+
```

Comment:
not checked

Reply:
or do you check it inside ehsm_parse_keyblob() ?

Reply:
it check inside ehsm_parse_keyblob() :
@line149 if (NULL == keyblob_data || NULL == plaintext || NULL == plaintext_size || plaintext_size < keyblob_data->ciphertext_size)
        return SGX_ERROR_INVALID_PARAMETER;